rust/lance-encoding/src/encoder.rs

```diff
@@ -133,6 +138,48 @@ impl Default for EncodedColumn {
     }
 }
 
+/// A tool to reserve space for buffers that are not in-line with the data
+///
+/// In most cases, buffers are stored in the page and referred to in the encoding
+/// metadata by their index in the page.  This keeps all buffers within a page together.
+/// As a result, most encoders should not need to use this structure.
+///
+/// In some cases (currently only the large binary encoding) there is a need to access
+/// buffers that are not in the page (becuase storing the position / offset of every page
+/// in the page metadata would be too expensive).
+///
+/// To do this you can add a buffer with `add_buffer` and then use the returned position
+/// in some way (in the large binary encoding the returned position is stored in the page
+/// data as a position / size array).
+pub struct OutOfLineBuffers {
+    position: u64,
+    buffers: Vec<LanceBuffer>,
+}
+
+impl OutOfLineBuffers {
+    pub fn new(base_position: u64) -> Self {
+        Self {
+            position: base_position,
+            buffers: Vec::new(),
+        }
+    }
+
+    pub fn add_buffer(&mut self, buffer: LanceBuffer) -> u64 {
+        let position = self.position;
+        self.position += buffer.len() as u64;
+        self.buffers.push(buffer);
+        position
+    }
+
+    pub fn take_buffers(&mut self) -> Vec<LanceBuffer> {
+        std::mem::take(&mut self.buffers)
+    }
```

Comment:
I wonder if it's safer to take ownership? Or do you still need the position afterwards?
```suggestion
    pub fn take_buffers(self) -> Vec<LanceBuffer> {
       self.buffers
    }
```

Reply:
Yes, the original goal was to avoid having to reobtain the position.  However, I think this is a relic from my initial approach and I can probably consume it.

My original approach tried to do the buffer reservations as part of the encode task but I couldn't come up with a good way to manage that since the encode tasks run in parallel and not necessarily in order.

Reply:
Switched the method to consume `self`.